pkg/telemetry/client.go

```diff
@@ -0,0 +1,274 @@
+	}
+
+	err := tc.metricsDB.CreateSchema()
+	if err != nil {
+		return errors.Wrap(err, "unable to create the telemetry schema")
+	}
+
+	return tc.metricsDB.SaveOperationMetric(tc.currentOperationMetrics)
+}
+
+// SendMetrics sends the local stored metrics to super collider
+// TODO: to be implemented
+func (tc *telemetryClient) SendMetrics() error {
+	return nil
+}
+
+func (tc *telemetryClient) isCoreCommand(cmdPath string) bool {
+	if cmdPath != "" {
+		cmds := strings.Split(cmdPath, " ")
+		if len(cmds) >= 2 && tc.coreCommandMap[cmds[1]] {
+			return true
+		}
+	}
+
+	return false
+}
+func (tc *telemetryClient) updateMetricsForCoreCommand(cmd *cobra.Command, args []string, cliID string) error {
+	tc.currentOperationMetrics.CliID = cliID
+	tc.currentOperationMetrics.CliVersion = buildinfo.Version
+	tc.currentOperationMetrics.StartTime = time.Now()
+	tc.currentOperationMetrics.CommandName = strings.Join(strings.Split(cmd.CommandPath(), " ")[1:], " ")
+
+	// CLI recommendation is to have a single argument for a command
+	if len(args) != 0 {
+		tc.currentOperationMetrics.NameArg = hashString(args[0])
+	}
+
+	flagMap := make(map[string]string)
+	hashRequired := isHashRequiredForCmdFlags(cmd.CommandPath())
+	cmd.Flags().Visit(func(flag *pflag.Flag) {
+		// capture the boolean and empty flag values as is
+		if !hashRequired || flag.Value.String() == "" || flag.Value.Type() == "bool" {
+			flagMap[flag.Name] = flag.Value.String()
+		} else {
+			flagMap[flag.Name] = hashString(flag.Value.String())
+		}
+	})
+	if len(flagMap) != 0 {
+		jsonString, _ := json.Marshal(flagMap)
+		tc.currentOperationMetrics.Flags = string(jsonString)
+	}
+
+	return nil
+}
+
+func (tc *telemetryClient) updateMetricsForPlugin(cmd *cobra.Command, args []string, cliID string) error {
+	tc.currentOperationMetrics.CliID = cliID
+	tc.currentOperationMetrics.CliVersion = buildinfo.Version
+	tc.currentOperationMetrics.StartTime = time.Now()
+
+	flagNames := TraverseFlagNames(args)
+	if len(flagNames) > 0 {
+		tc.currentOperationMetrics.Flags = flagNamesToJSONString(flagNames)
+	}
+
+	plugin := tc.pluginInfoFromCommand(cmd)
+	if plugin != nil {
+		tc.currentOperationMetrics.PluginName = plugin.Name
+		tc.currentOperationMetrics.PluginVersion = plugin.Version
+		tc.currentOperationMetrics.Target = string(plugin.Target)
+		tc.currentOperationMetrics.Endpoint = getEndpointSHA(plugin)
+	}
+
+	// TODO : Fix the command chain for plugins by using command chain cache constructed from generate-all-docs command
+	tc.currentOperationMetrics.CommandName = strings.Join(strings.Split(cmd.CommandPath(), " ")[1:], " ")
+
+	return nil
+}
+func (tc *telemetryClient) pluginInfoFromCommand(cmd *cobra.Command) *cli.PluginInfo {
+	var plugin *cli.PluginInfo
+
+	cmds := strings.Split(cmd.CommandPath(), " ")
+
+	// check if the second command in command path matches any target,
+	// if so, look for the 3rd command in the plugins map to find the plugin
+	cmdTarget := configtypes.StringToTarget(cmds[1])
```

Comment:
I have similar thoughts here as https://github.com/vmware-tanzu/tanzu-cli/pull/391/files#r1260607600
Do we need to make the assumptions we make here about the command path in order to find the plugin involved (if it is a plugin-based command)
Conceptually, being able to map each plugin-based Command back to the corresponding plugininfo would seem a more direct (and more future-proof) way to implement this function

Reply:
Thanks! 
Same as above comment. I would be using the Command object (a new annotation `pluginInstallationPath`:<plugin/installation/path>` would be added and would use this as key to find the plugin) used for the plugin command to find the plugin involved.
Updated the PR.